sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala

```diff
@@ -118,19 +118,32 @@ private[hive] object IsolatedClientLoader extends Logging {
       hadoopVersion: String,
       ivyPath: Option[String],
       remoteRepos: String): Seq[URL] = {
+    val hadoopJarNames = if (hadoopVersion.startsWith("3")) {
+      Seq(s"org.apache.hadoop:hadoop-client-api:$hadoopVersion",
+        s"org.apache.hadoop:hadoop-client-runtime:$hadoopVersion")
+    } else {
+      Seq(s"org.apache.hadoop:hadoop-client:$hadoopVersion")
+    }
     val hiveArtifacts = version.extraDeps ++
       Seq("hive-metastore", "hive-exec", "hive-common", "hive-serde")
         .map(a => s"org.apache.hive:$a:${version.fullVersion}") ++
-      Seq("com.google.guava:guava:14.0.1",
-        s"org.apache.hadoop:hadoop-client:$hadoopVersion")
+      Seq("com.google.guava:guava:14.0.1") ++ hadoopJarNames
+
+    val extraExclusions = if (hadoopVersion.startsWith("3")) {
+      // this introduced from lower version of Hive could conflict with jars in Hadoop 3.2+, so
+      // exclude here in favor of the ones in Hadoop 3.2+
+      Seq("org.apache.hadoop:hadoop-auth")
```

Comment:
Why is `hadoop-auth` the only one excluded here? Won't Hive pull in a variety of the non-shaded Hadoop JARs?

Reply:
I only excluded this one because it caused issues, but yes we could exclude more.

A bit more context, I think currently the way it works is a bit messed up: it pulls hadoop dependencies from both Hive and the explicit `hadoop-client` dependency, while the latter uses the Hadoop version in Spark (which is 3.x in default, as opposed to 2.x in Hive). As result, one can end up with two versions of various Hadoop jars in the class path and conflicts can happen unexpectedly. This only happens when we are not sharing Hadoop classes though (which is rare): in the other path Spark always loads its own Hadoop classes first, then Hive's.

Reply:
I thought that Ivy would do dependency conflict resolution to resolve such issues? Do you actually get two versions of the same module (e.g. `hadoop-common` 2.7 and 3.2) or mixed Hadoop modules (like `hadoop-common` 2.7 but `hadoop-auth` 3.2)? I was under the impression that the former should be resolved automatically.

In any case -- this seems like a good opportunity to clean it up. If we exclude everything _except_ `hadoop-client-api` and `hadoop-client-runtime`, that should cover excluding all of Hive's JARs, right?

Reply:
No not same artifacts with different versions, but different artifacts with different versions (e.g., `hadoop-common` 2.7 but `hadoop-auth` 3.2).

Yes we could do a cleanup and remove all other dependencies here, although somewhat related I'm wondering whether we do need to support not-sharing classes at all. It is currently used only in tests and in a rare case where it can't find [the specified hadoop version](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L78) (I don't fully understand why it switch to not using shared classes in the latter).

Reply:
It also related to the definition of "not sharing Hadoop classes": does it mean Hadoop classes from Hive dependency take precedence over those from Spark itself? if that's true then we probably shouldn't exclude those.

Reply:
True... I've recently been debugging some test issues in our environment that are caused by this behavior, and am also confused by its existence.

Given that we explicitly specify a Hadoop version to `downloadVersion` (which can be different from the one the rest of Spark is using), it seems that should take precedence over the ones Hive pulls in.

Reply:
Yes it's confusing. I'll probably leave this as it is and open another JIRA specifically for this issue, and see what the broader community think.